### ISSUE TITLE
B031 false positive in loop with continue/break/return

### ISSUE THREAD

**Issue by MikkelSchubert** [astral-sh/ruff#10337]
```python
from __future__ import annotations
from itertools import groupby

def ruff_b031_false_positive(values: list[int]) -> None:
    for key, group in groupby(values):
        if key:
            list(group)
            continue # or break or return

        list(group)
```

```console
$ ruff check --isolated --select B031 testcase.py
testcase.py:10:14: B031 Using the generator returned from `itertools.groupby()` more than once will do nothing on the second usage
Found 1 error.
```

There have been fixes for when the generator is used in different branches of an `if` statements, see for example #4050, but the warning still triggers if you use `continue`, `break`, or `return`. Tested with Ruff version 0.3.2.

**Comment by dhruvmanila** [comment]
Thanks! 

I think this can be solved. We would need to consider the block outside the `if` statement as if it's mutually exclusive with the `if` block if there's a `continue` / `break` / `return` as the last statement in the `if` block.

(We so need control flow graphs :))

**Comment by yt2b** [comment]
@dhruvmanila 
Can I try to work on this?

**Comment by dhruvmanila** [comment]
@yt2b Sure! It might be a bit tricky to get it right because we don't have control flow graph and the implementation uses visitor and stack to simulate it. This [comment](https://github.com/astral-sh/ruff/pull/3844#issuecomment-1495229979) has a visualization on what it looks like. Feel free to ping me with any questions you might have :)

**Comment by dhruvmanila** [comment]
(Assigning to you but don't feel any pressure to complete it.)
